packages/synapse-guest-module/synapse_guest_module/guest_module.py

```diff
@@ -138,6 +141,34 @@ async def callback_user_may_invite(
         user_is_guest = inviter.startswith("@" + self._config.user_id_prefix)
         return not user_is_guest
 
+    async def callback_user_may_join_room(
+        self,
+        user_id: str,
+        room_id: str,
+        is_invited: bool
+    ) -> str:
+        """Returns whether this user is allowed to join a room. Guest users
+        should only be able to do that if the room is Ask to Join (knock).
+        """
+        user_is_guest = user_id.startswith("@" + self._config.user_id_prefix)
+        if not user_is_guest or is_invited:
+            return NOT_SPAM
+
+        join_rules_events = await self._api.get_state_events_in_room(
+            room_id,
+            [("m.room.join_rules", None)]
+        )
+        if join_rules_events is None or len(join_rules_events) == 0:
+            return errors.Codes.BAD_STATE
+
+        for event in join_rules_events:
+            join_rule = event.get("content", {})
+            is_knock = join_rule.get("join_rule") == "knock"
```

Comment:
There is also knock_restricted, not sure how that is even exposed through element's UI; it might not be

Reply:
Don't think it is but I've updated it to check for "knock*"